Provide Python 3.8.0a4

h-vetinari opened the pull request:
Next step for #254, following the work of @jjhelmus in #258.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
@conda-forge-admin, please rerender

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to re-render for you, but it looks like there was nothing to do.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
> @mingwandroid: Any chance of not attributing patches I wrote to yourself?
I don't want people to query you, instead of me, about them.

Absolutely - that patch was a c&p artefact (generated by my own git)... Will double-check and repush shortly.

h-vetinari commented:
@mingwandroid
This is my first time patching git patches, so there were some growing pains - contributions are fixed again.

I now also tried to religiously keep the original line endings intact, leaving substantial changes only in patch [0003](https://github.com/conda-forge/python-feedstock/blob/3fcf73efdae75e2b462c3bb396c79e2c522baf79/recipe/patches/0003-Win32-distutils-Also-look-for-executable.bat.patch). The file [Lib/distutils/spawn.py](https://github.com/python/cpython/blob/c1004b85464579771db3b50f5b150449275bbbd0/Lib/distutils/spawn.py) has changed a fair bit from `3.8.0a3`, so please double-check that I'm doing the right thing here.

In patch [0007](https://github.com/conda-forge/python-feedstock/blob/3fcf73efdae75e2b462c3bb396c79e2c522baf79/recipe/patches/0007-Support-cross-compiling-byte-code.patch), I made sure that every `$(PYTHON_FOR_BUILD)` is preceded by a `$(PY_BUILD_ENVIRON)` -- there were a few that didn't have it so far.

In patch [0020](https://github.com/conda-forge/python-feedstock/blob/3fcf73efdae75e2b462c3bb396c79e2c522baf79/recipe/patches/0020-Add-CondaEcosystemModifyDllSearchPath.patch), I'd like to ask you to check if I'm adding the code at the right point in the [file](https://github.com/python/cpython/blob/c1004b85464579771db3b50f5b150449275bbbd0/Modules/main.c)

The builds are still failing unfortunately, although I haven't debugged the [following](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=34080&view=logs) yet:
`linux linux_target_platformlinux-64`:
<details>

```
~/feedstock_root/build_artifacts/python_1557517148256/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac ~/feedstock_root/build_artifacts/python_1557517148256/work
++ find 'lib/python3.8/config-3.8m*' -name libpython3.8m.a
find: `lib/python3.8/config-3.8m*': No such file or directory
+ CONFIG_LIBPYTHON=
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 447, in execute
    verify=args.verify, variants=args.variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 208, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2314, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1477, in build
    cwd=src_dir, stats=build_stats)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 374, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 354, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-e', '/home/conda/feedstock_root/build_artifacts/python_1557517148256/work/conda_build.sh']' returned non-zero exit status 1.
##[error]Bash exited with code '1'.
##[section]Finishing: Run docker build
```

</details>

`osx osx_target_platformosx-64`:

<details> 

```
+ pushd /usr/local/miniconda/conda-bld/python_1557517159253/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh
+ [[ -f lib/libpython3.8m.a ]]
$SRC_DIR
$PREFIX $SRC_DIR
++ find 'lib/python3.8/config-3.8m*' -name libpython3.8m.a
find: lib/python3.8/config-3.8m*: No such file or directory
+ CONFIG_LIBPYTHON=
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 456, in main
    execute(sys.argv[1:])
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 447, in execute
    verify=args.verify, variants=args.variants)
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/api.py", line 208, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/build.py", line 2314, in build_tree
    notest=notest,
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/build.py", line 1477, in build
    cwd=src_dir, stats=build_stats)
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/utils.py", line 374, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/usr/local/miniconda/lib/python3.7/site-packages/conda_build/utils.py", line 354, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-e', '/usr/local/miniconda/conda-bld/python_1557517159253/work/conda_build.sh']' returned non-zero exit status 1.
##[error]Bash exited with code '1'.
##[section]Finishing: Build recipe
```

</details>

The windows build fails with the same error as in #258, which wasn't [an issue](https://github.com/conda-forge/python-feedstock/pull/258#issuecomment-485583191) apparently:

<details>

```
Build started 5/10/2019 7:41:41 PM.
     1>Project "%SRC_DIR%\PCbuild\pcbuild.proj" on node 1 (Build target(s)).
     1>Project "%SRC_DIR%\PCbuild\pcbuild.proj" (1) is building "%SRC_DIR%\PCbuild\pythoncore.vcxproj" (2) on node 1 (Build target(s)).
     2>%SRC_DIR%\PCbuild\python.props(97,31): error MSB4086: A numeric comparison was attempted on "$([System.Version]::Parse($(_RegistryVersion)))" that evaluates to "10.0.17763.0" instead of a number, in condition "$([System.Version]::Parse($(_RegistryVersion))) > $([System.Version]::Parse($(DefaultWindowsSDKVersion)))". [%SRC_DIR%\PCbuild\pythoncore.vcxproj]
     2>Done Building Project "%SRC_DIR%\PCbuild\pythoncore.vcxproj" (Build target(s)) -- FAILED.
     1>Done Building Project "%SRC_DIR%\PCbuild\pcbuild.proj" (Build target(s)) -- FAILED.

Build FAILED.

       "%SRC_DIR%\PCbuild\pcbuild.proj" (Build target) (1) ->
       "%SRC_DIR%\PCbuild\pythoncore.vcxproj" (Build target) (2) ->
         %SRC_DIR%\PCbuild\python.props(97,31): error MSB4086: A numeric comparison was attempted on "$([System.Version]::Parse($(_RegistryVersion)))" that evaluates to "10.0.17763.0" instead of a number, in condition "$([System.Version]::Parse($(_RegistryVersion))) > $([System.Version]::Parse($(DefaultWindowsSDKVersion)))". [%SRC_DIR%\PCbuild\pythoncore.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.66
Traceback (most recent call last):
```

</details>

jakirkham commented on .gitattributes:
```diff
@@ -1,6 +1,5 @@
 * text=auto
 
-*.patch binary
```
This should be kept.

h-vetinari replied:
To be honest, I thought this was an artefact from times long past. Github already ignores it anyway, and I'd also like to be able to see the diff in my git locally before committing. What's the reason for keeping it?

Not that I care much about this, I can carry a local diff as well.

mingwandroid replied:
This is not some left over junk. The python source code contains files that contain Windows line endings, so the patches to those files must also contain Windows line endings.

h-vetinari replied:
@mingwandroid 
The patches applied fine without marking them as binary. It's a pain to keep the right line endings when editing files that use a different convention (regardless of whether they're marked "binary"), but it seems to me that the only difference is whether I can see a git diff locally or not.

Of course, I may be overlooking something else, and so I reverted that change.

mingwandroid replied:
Yes you are. Git has the concept of "line ending check out mode", unfortunately. If we aren't explicit about the lines endings then people who check out the repo in lf mode well be unable to apply the patches. 

You did say you were new to git? I am far from new to it.

h-vetinari replied:
> You did say you were new to git? I am far from new to it.

New to editing patch files by hand, yes; new to git, no.

Fair enough on the line endings. Guess my defaults took care of that.

In any case, I reverted it already. Any tips on the failures? Did you have a chance to review the patches I mentioned above?

mingwandroid replied:
> New to editing patch files by hand, yes; new to git, no.

To work with theses patches,  I use a clone of the upstream git repo. I would never work on them by hand instead using git am,  rebase and format-patch.

They looked reasonable upon eyeballing and thanks for taking care there. I can't help or with the failures at present, sorry, I'm too busy.

h-vetinari replied:
> To work with theses patches, I use a clone of the upstream git repo. I would never work on them by hand instead using git am, rebase and format-patch.

I was using git am and `git apply -v` to explore, and `format-patch` to generate new patches, but then pasting things into your patches so I don't change attribution. But yeah, I can see how using the upstream repo would help a lot there. :)

> I can't help or with the failures at present, sorry, I'm too busy.

OK, I'll try to have a look when I have more time.

jakirkham replied:
@h-vetinari, this file is autogenerated. The change will just be overwritten anyways. It would be good to revert it to it's previous autogenerated state. If you find issue with the contents of this file, that should be raised over at [conda-smithy]( https://github.com/conda-forge/conda-smithy ) not in individual feedstocks. Thanks for your help. 🙂

h-vetinari commented:
> @jakirkham: @h-vetinari, this file is autogenerated. The change will just be overwritten anyways. It would be good to revert it to it's previous autogenerated state. If you find issue with the contents of this file, that should be raised over at conda-smithy not in individual feedstocks. Thanks for your help. 🙂

Thanks for the tip. Following my experience above, it may be that things work without marking patches as binary in a recent git, but I don't have time to test whether that's actually cross-platform, or which git-version is the necessary minimum. I think it's ok to leave for now.

For reference: conda-forge/conda-smithy#549 conda-forge/conda-smithy#550.

mingwandroid commented:
> Following my experience above, it may be that things work without marking patches as binary in a recent git

How so? The patches are applied with patch, not git and they are applied to a tarball not a git repo.

Where is git involved? It's only the tool I use to *work* on these patches out of personal preference (and I have to be careful about crlf, you cannot just flatten a patch to one type or the other, a patch can contain blocks referring to different files with different line endings (and if course those can even change within a file so a patch containing mixed crlf and lf can be necessary even when it touches a single file..  In fact patch headers are always lf so every crlf patch is mixed really)).

The fact is the patches much be correct wrt crlf mode on everyone's computer from a git clone irrespective of the gitconfig settings and the only way to guarantee that with git is to mark them as binary.

If anyone knows better then I am happy to be corrected!

h-vetinari commented:
@conda-forge-admin, please rerender

h-vetinari commented:
Tried to address a TODO comment in meta.yaml (moving some packages from external to build/run-dependencies), but apparently, I didn't understand the full impact of that. Reverted the corresponding commit.

Also, I'm struggling to figure out the build failures. I added a comment in `build.sh` where the following happens:
```
[...]
++ find 'lib/python3.8/config-3.8m*' -name libpython3.8m.a
find: `lib/python3.8/config-3.8m*': No such file or directory
+ CONFIG_LIBPYTHON=
Traceback (most recent call last):
[...]
```

Trying to pick this up again since 3.8b1 is around the corner, and the necessary changes will probably be easier if 3.8a4 is already through the gate... @jjhelmus @mingwandroid @jakirkham

alimcmaster1 commented:
That failure is also not obvious to me either.. ( I can potentially have a more detailed look over the weekend ) Maintainers any ideas -> we are hoping to get 3.8.0a4 into conda forge so open source projects that use conda-forge channel can test against python3.8 - any help would be much appreciated. @jjhelmus @mingwandroid @jakirkham

h-vetinari commented:
I took the time to try to run `conda build recipe` on an ubuntu 18.04 machine and I did not get the same error (however, after building python, the test-suite was executed and failed at `test_concurrent_futures`).

h-vetinari commented:
@conda-forge-admin, please rerender

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to re-render for you but ran into some issues, please ping conda-forge/core for further assistance.

h-vetinari commented:
Closing as superseded by #268 / #271.
